### PR TITLE
Animation3D

### PR DESCRIPTION
**Pull request by super626**
We are keeping on testing this, so just comment it.

c3b support
improve performance. shader simplify, use VAO and batch draw
animation3d fix, load original pose, transform bugs on bones without skin.
more test on animate3d, we are keeping on update this.

### CODE REVIEW

File: cocos/3d/CCAnimation3D.h [hunk 1]
```diff
@@ -43,31 +43,39 @@ class Animation3D: public Ref
...
         Curve();
         ~Curve();
     };
     
-    //read all animation or only the animation with given animationName? animationName == "" read all.
+    /**read all animation or only the animation with given animationName? animationName == "" read all.*/
     static Animation3D* getOrCreate(const std::string& filename, const std::string& animationName = "");
     
+//    /**read all animation or only the animation with given animationName? animationName == "" read all.*/
+//    static Animation3D* create(const std::string& filename, const std::string& animationName = "");
```
Comment (minggo) [inline cocos/3d/CCAnimation3D.h]: Please remove unneeded codes.

File: cocos/3d/CCBundleReader.h [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+    /** creates an BundleReader with lpbuffer and length 
+     * @param lpbuffer The pointer to the file data 
+     * @param length The size for lpbuffer in bytes
+     */
+    //static BundleReader* create(char* lpbuffer, unsigned int length);
+
+    /**
+     * Close and delete buffer
+	 */
```
Comment (minggo) [inline cocos/3d/CCBundleReader.h]: Wrong indent.
Many places in this file.

File: cocos/3d/CCBundleReader.h [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+     * Reads an array of elements.
+	 *
+	 * @param ptr   The pointer to the memory to copy into.
+	 *              The available size should be at least bytes.
+	 * @param size  The size of each element to be read, in bytes.
+	 * @param count The number of elements to read.
+	 *
+	 * @return The number of elements read.
+     */
+    size_t read(void* ptr, size_t size, size_t count);
```
Comment (minggo) [inline cocos/3d/CCBundleReader.h]: We suggested using `ssize_t` instead of `size_t`.

File: cocos/3d/CCMeshSkin.h [hunk 4]
```diff
@@ -72,24 +81,42 @@ class Bone : public Ref
...
+     * @param m Mat4 representing the inverse bind pose for this Bone.
      */
     void setInverseBindPose(const Mat4& m);
     
     /**
+     * Sets the bone's original pose.
+     *
+     * @param m Mat4 representing the original pose for this Bone.
+     */
+    void setOriPose(const Mat4& m);
```
Comment (minggo) [inline cocos/3d/CCMeshSkin.h:93]: Is it a common way to use `Ori` short of `Original`?

**Comment by minggo** [comment]
Please use small picture.

File: cocos/3d/CCAnimation3D.h (hunk 1 above)
Comment (super626) [inline cocos/3d/CCAnimation3D.h]: Ok, thanks

File: cocos/3d/CCBundleReader.h (hunk 3 above)
Comment (super626) [inline cocos/3d/CCBundleReader.h]: Ok, thanks

File: cocos/3d/CCBundleReader.h (hunk 2 above)
Comment (super626) [inline cocos/3d/CCBundleReader.h]: Ok, thanks

File: cocos/3d/CCMeshSkin.h (hunk 4 above)
Comment (super626) [inline cocos/3d/CCMeshSkin.h:93]: 3D usually use Ori for short.
Is it common for cocos2d-x? I will change it if not.

Comment (minggo) [inline cocos/3d/CCMeshSkin.h:93]: I am not sure, just a question.
If it is a known to use it for short.
Then, not problem.

**Pull request merged**